docs/heap_profiling.md

```diff
@@ -0,0 +1,140 @@
+# Heap Profiling
+
+The JAX Heap Profiler allows us to explore how and why JAX programs are using
+GPU or TPU memory. For example, it can be used to:
+
+* Figure out which arrays and executables are in GPU memory at a given time, or
+* Track down memory leaks.
+
+## Installation
+
+The JAX heap profiler emits output that can be interpreted using
+pprof (<https://github.com/google/pprof>). Start by installing `pprof`,
+by following its
+[installation instructions](https://github.com/google/pprof#building-pprof).
+At the time of writing, installing `pprof` requires first installing
+[Go](https://golang.org/) and [Graphviz](http://www.graphviz.org/), and then
+running
+
+```shell
+go get -u github.com/google/pprof
+```
+
+which installs `pprof` as `$GOPATH/bin/pprof`, where `GOPATH` defaults to
+`~/go`.
+
+```{note}
+The version of `pprof` from <https://github.com/google/pprof> is not the same as
+the older tool of the same name distributed as part of the `gperftools` package.
+The `gperftools` version of `pprof` will not work with JAX.
+```
+
+## Understanding how a JAX program is using GPU or TPU memory
+
+A common use of the heap profiler is to figure out why a JAX program is using
+a large amount of GPU or TPU memory, for example if trying to debug an
+out-of-memory problem.
+
+To capture a heap profile to disk, use
+{func}`jax.profiler.save_heap_profile`. For example, consider the following
+Python program:
+
+```python
+import jax
+import jax.numpy as jnp
+import jax.profiler
+
+def func1(x):
+  return jnp.tile(x, 10) * 0.5
+
+def func2(x):
+  y = func1(x)
+  return y, jnp.tile(x, 10) + 1
+
+x = jax.random.normal(jax.random.PRNGKey(42), (1000, 1000))
+y, z = func2(x)
+
+z.block_until_ready()
+
+jax.profiler.save_heap_profile("heap.prof")
+```
+
+If we first run the program above and then execute
+
+```shell
+pprof --web heap.prof
+```
+
+`pprof` opens a web browser containing the following visualization of the heap
+profile in callgraph format:
+
+![Heap profiling example](_static/heap_profile.svg)
+
+The pprof documentation explains
+[how to interpret callgraph visualizations](https://github.com/google/pprof/blob/master/doc/README.md#interpreting-the-callgraph).
```

Comment:
nit: maybe get rid of this, and say something like this at the end of the next paragraph: "For more information on interpreting the callgraph visualizations, see [the pprof documentation](...)"

When I read this, I at first thought this was all that was gonna be said about the visualization.